### PR TITLE
[dapp-kit] Add useNetworkConfig hook for managing arbitrary network c…

### PR DESCRIPTION
**Pull request by hayes-mysten**
…onfigs

## Description 

Describe the changes or additions included in this PR.

## Test Plan 

How did you test the new or updated feature?

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: sdk/docs/pages/dapp-kit/sui-client-provider.mdx [hunk 1]
```diff
@@ -140,3 +140,52 @@ function NetworkSelector() {
...
+					<YourApp />
+				</WalletProvider>
+			</SuiClientProvider>
+		</QueryClientProvider>
+	);
+}
+
+
+function YourApp() {
+	const { myMovePackageId } = useSuiClientContext();
```
Comment (williamrobertson13) [inline sdk/docs/pages/dapp-kit/sui-client-provider.mdx:187]: I find these docs a bit confusing - above we say we can use `useNetworkConfig` to retrieve the network config information, but we're using `useSuiClientContext` here. It also isn't obvious to me from the docs that I need to define and export my `createNetworkConfig` result in a separate file if I'd like to use the hook around the codebase. In the example above, we just create it with the top-level App definition and then never use it

File: sdk/dapp-kit/src/hooks/networkConfig.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import type { SuiClientOptions } from '@mysten/sui.js/client';
+
+import { useSuiClientContext } from './useSuiClient.js';
+
+export function createNetworkConfigs<T extends Record<string, SuiClientOptions>>(
+	networkConfigs: T,
+): { networkConfigs: T; useNetworkConfig: () => T[keyof T] } {
+	return {
+		networkConfigs,
+		useNetworkConfig: () => {
```
Comment (williamrobertson13) [inline sdk/dapp-kit/src/hooks/networkConfig.ts:13]: I might just not fully understand, but what is the purpose of having this function return a hook versus just having `useNetworkConfig` readily available? I'm guessing it's so we get the strong typing with the `as T[keyof T]` cast?

Comment (hayes-mysten) [reply sdk/dapp-kit/src/hooks/networkConfig.ts]: Yes, it's basically just for type-safety.  I personally like this better than making someone pass type arguments

File: sdk/docs/pages/dapp-kit/sui-client-provider.mdx [hunk 3]
```diff
@@ -140,3 +140,52 @@ function NetworkSelector() {
...
+
+// Config options for the networks you want to connect to
+const { networkConfigs, useNetworkConfig } = createNetworkConfigs({
+	localnet: {
+		url: getFullnodeUrl('localnet')
+		myMovePackageId: '0x123',
+	},
+	mainnet: {
+		url: getFullnodeUrl('mainnet')
+		myMovePackageId: '0x456',
```
[the remark this answers is not in the document]
Comment (hayes-mysten) [reply sdk/docs/pages/dapp-kit/sui-client-provider.mdx]: I think we just leave this up to the app.  The current implementation lets you do either without any issues

File: sdk/dapp-kit/src/hooks/networkConfig.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+export type NetworkConfig<T extends object = object> = SuiClientOptions & {
+	variables?: T;
+};
+
+export function createNetworkConfigs<
+	const T extends Record<string, Config>,
+	Config extends NetworkConfig<Variables> = T[keyof T],
+	Variables extends object = NonNullable<Config['variables']>,
+>(networkConfigs: T) {
+	return {
```
Comment (Jordan-Mysten) [inline sdk/dapp-kit/src/hooks/networkConfig.ts:17]: I hate depending on function hoisting, this hurts my brain

**Review by Jordan-Mysten — COMMENTED** [review]
Overal LGTM, two small thoughts:
- I think it's a bummer that our general solution here requires React, because it means things like the Kiosk SDK (which basically does this, just internally) can't use the primitive here. Not something we need to solve now, but as we move into things like code / library gen, stable package IDs, and .move resolution, we should try to find a way to make this work in lower-level libraries as well as in Dapp Kit.
- I wonder if `networkConfigs` should be `networkConfig`

File: sdk/dapp-kit/src/hooks/networkConfig.ts (hunk 4 above)
Comment (hayes-mysten) [reply sdk/dapp-kit/src/hooks/networkConfig.ts:17]: I love function hoisting... it makes things so much easier for me to read, but I can change it if you want.

with this setup its:

1. Inputs
2. Return types
3. the implementation which you probably don't care about

Comment (Jordan-Mysten) [reply sdk/dapp-kit/src/hooks/networkConfig.ts:17]: The implementation is the _only_ thing I care about.

Comment (Jordan-Mysten) [reply sdk/dapp-kit/src/hooks/networkConfig.ts:17]: More seriously, I just think people these days don't understand it, especially with `var` being dead and nothing else getting hoisted.

**Review by williamrobertson13 — APPROVED** [review]

**Pull request merged**